File: backend/app/services/cost.py
"""日均成本计算。"""
from dataclasses import dataclass
from datetime import date

from app.models import Asset, AssetStatus


@dataclass
class CostInfo:
    period_days: int
    total_cost: float
    daily_cost: float
    formula: str


def calc_cost(asset: Asset, today: date) -> CostInfo:
    """按状态计算日均成本。

    - 使用中: 购买价 / 自购买起天数；勾选「到期日期」且有到期日时按 (到期日 - 购买日)
    - 已售出: (购买价 - 售出价) / (售出日 - 购买日)
    - 已损坏: 购买价 / (损坏日 - 购买日)
    - 已过期: 购买价 / (到期日 - 购买日)
    """
    formula = AssetStatus.in_use.value
    total_cost = asset.purchase_price

    if asset.status == AssetStatus.sold.value and asset.sale_date and asset.sale_price is not None:
        days = (asset.sale_date - asset.purchase_date).days
        total_cost = round(asset.purchase_price - asset.sale_price, 2)
        formula = AssetStatus.sold.value
    elif asset.status == AssetStatus.broken.value and asset.broken_date:
        days = (asset.broken_date - asset.purchase_date).days
        formula = AssetStatus.broken.value
    elif asset.status == AssetStatus.expired.value and asset.expiry_date:
        days = (asset.expiry_date - asset.purchase_date).days
        formula = AssetStatus.expired.value
    elif (
        asset.status == AssetStatus.in_use.value
        and asset.expiry_date is not None
        and asset.category is not None
        and asset.category.has_expiry
    ):
        days = (asset.expiry_date - asset.purchase_date).days
        formula = AssetStatus.expired.value
    else:
        days = (today - asset.purchase_date).days
        formula = AssetStatus.in_use.value

    daily_cost = round(total_cost / days, 4) if days > 0 else 0.0
    return CostInfo(period_days=max(days, 0), total_cost=total_cost, daily_cost=daily_cost, formula=formula)


def sync_expiry_status(asset: Asset, today: date) -> bool:
    """勾选「到期日期」类别的资产，按到期日自动同步状态：过期则标记已过期，日期改到未来则恢复使用中。"""
    if asset.category is None or not asset.category.has_expiry:
        return False
    if asset.status == AssetStatus.in_use.value and asset.expiry_date is not None and asset.expiry_date < today:
        asset.status = AssetStatus.expired.value
        return True
    if asset.status == AssetStatus.expired.value and (asset.expiry_date is None or asset.expiry_date >= today):
        asset.status = AssetStatus.in_use.value
        return True
    return False


def reminder_target_dates(asset: Asset, today: date | None = None) -> list[tuple[str, date]]:
    """返回资产有效的提醒目标日期列表 [(date_type, target_date), ...]
    date_type 为 'warranty' 或 'expiry'。
    如果提供了 today，只返回 target_date >= today 的未来目标。
    """
    targets: list[tuple[str, date]] = []
    if asset.warranty_end_date and (today is None or asset.warranty_end_date >= today):
        targets.append(("warranty", asset.warranty_end_date))
    if asset.expiry_date and (today is None or asset.expiry_date >= today):
        targets.append(("expiry", asset.expiry_date))
    targets.sort(key=lambda t: t[1])
    return targets


def reminder_base_date(asset: Asset, today: date | None = None) -> date | None:
    """提醒基准日期：若传入 today 则优先返回最近的未过期目标日期（保修或到期）；未传入则优先保修。"""
    targets = reminder_target_dates(asset, today)
    if targets:
        return targets[0][1]
    return asset.warranty_end_date or asset.expiry_date

File: backend/internal/services/cost.go
package services

import (
	"math"
	"sort"
	"time"

	"thingspan/internal/models"
)

func ParseDate(s string) (time.Time, error) {
	return time.Parse("2006-01-02", s)
}

func DaysBetween(startDate, endDate string) int {
	t1, err1 := ParseDate(startDate)
	t2, err2 := ParseDate(endDate)
	if err1 != nil || err2 != nil {
		return 0
	}
	// Use UTC midnight to calculate integer calendar days accurately
	t1 = time.Date(t1.Year(), t1.Month(), t1.Day(), 0, 0, 0, 0, time.UTC)
	t2 = time.Date(t2.Year(), t2.Month(), t2.Day(), 0, 0, 0, 0, time.UTC)
	return int(t2.Sub(t1) / (24 * time.Hour))
}

func AddDays(startDate string, days int) string {
	t, err := ParseDate(startDate)
	if err != nil {
		return ""
	}
	return t.AddDate(0, 0, days).Format("2006-01-02")
}

func RoundFloat(val float64, precision int) float64 {
	p := math.Pow10(precision)
	return math.Round(val*p) / p
}

// CalcCost computes the daily cost and period based on asset status and dates
func CalcCost(asset *models.Asset, today string) *models.CostOut {
	formula := models.StatusInUse
	totalCost := asset.PurchasePrice
	days := 0

	if asset.Status == models.StatusSold && asset.SaleDate != nil && *asset.SaleDate != "" && asset.SalePrice != nil {
		days = DaysBetween(asset.PurchaseDate, *asset.SaleDate)
		totalCost = RoundFloat(asset.PurchasePrice-*asset.SalePrice, 2)
		formula = models.StatusSold
	} else if asset.Status == models.StatusBroken && asset.BrokenDate != nil && *asset.BrokenDate != "" {
		days = DaysBetween(asset.PurchaseDate, *asset.BrokenDate)
		formula = models.StatusBroken
	} else if asset.Status == models.StatusExpired && asset.ExpiryDate != nil && *asset.ExpiryDate != "" {
		days = DaysBetween(asset.PurchaseDate, *asset.ExpiryDate)
		formula = models.StatusExpired
	} else if asset.Status == models.StatusInUse &&
		asset.ExpiryDate != nil && *asset.ExpiryDate != "" &&
		asset.Category != nil && asset.Category.HasExpiry {
		days = DaysBetween(asset.PurchaseDate, *asset.ExpiryDate)
		formula = models.StatusExpired
	} else {
		days = DaysBetween(asset.PurchaseDate, today)
		formula = models.StatusInUse
	}

	dailyCost := 0.0
	if days > 0 {
		dailyCost = RoundFloat(totalCost/float64(days), 4)
	}

	periodDays := days
	if periodDays < 0 {
		periodDays = 0
	}

	return &models.CostOut{
		PeriodDays: periodDays,
		TotalCost:  totalCost,
		DailyCost:  dailyCost,
		Formula:    formula,
	}
}

// SyncExpiryStatus updates asset status dynamically between in_use and expired
func SyncExpiryStatus(asset *models.Asset, today string) bool {
	if asset.Category == nil || !asset.Category.HasExpiry {
		return false
	}

	hasExpiryDate := asset.ExpiryDate != nil && *asset.ExpiryDate != ""
	if asset.Status == models.StatusInUse && hasExpiryDate && *asset.ExpiryDate < today {
		asset.Status = models.StatusExpired
		return true
	}
	if asset.Status == models.StatusExpired && (!hasExpiryDate || *asset.ExpiryDate >= today) {
		asset.Status = models.StatusInUse
		return true
	}

	return false
}

// ApplyWarranty calculates warranty_end_date = purchase_date + (warranty_months * 30 days)
func ApplyWarranty(asset *models.Asset, category *models.Category, force bool) {
	hasWarrantyMonths := asset.WarrantyMonths != nil && *asset.WarrantyMonths > 0
	if category != nil && category.HasWarranty && asset.PurchaseDate != "" && hasWarrantyMonths {
		if force || asset.WarrantyEndDate == nil || *asset.WarrantyEndDate == "" {
			endDate := AddDays(asset.PurchaseDate, *asset.WarrantyMonths*30)
			asset.WarrantyEndDate = &endDate
		}
	} else if force {
		asset.WarrantyEndDate = nil
	}
}

type ReminderTarget struct {
	DateType   string // "warranty" | "expiry"
	TargetDate string // YYYY-MM-DD
}

// ReminderTargetDates returns list of targets (warranty and expiry), optionally filtered by target_date >= today
func ReminderTargetDates(asset *models.Asset, today string) []ReminderTarget {
	var targets []ReminderTarget
	if asset.WarrantyEndDate != nil && *asset.WarrantyEndDate != "" {
		if today == "" || *asset.WarrantyEndDate >= today {
			targets = append(targets, ReminderTarget{
				DateType:   "warranty",
				TargetDate: *asset.WarrantyEndDate,
			})
		}
	}
	if asset.ExpiryDate != nil && *asset.ExpiryDate != "" {
		if today == "" || *asset.ExpiryDate >= today {
			targets = append(targets, ReminderTarget{
				DateType:   "expiry",
				TargetDate: *asset.ExpiryDate,
			})
		}
	}
	sort.Slice(targets, func(i, j int) bool {
		return targets[i].TargetDate < targets[j].TargetDate
	})
	return targets
}

